src/include/api.h

```diff
@@ -55,6 +55,36 @@
 #define WT_SINGLE_THREAD_CHECK_STOP(s)
 #endif
 
+/*
+ * Helper macros to correctly read and check the API counters. Carefully read the out counter before
+ * the in counter otherwise the out counter can include more API calls than the in and make the
+ * balance negative.
+ */
+#define WT_API_COUNTER_REALIZE(s, counter, output)                                      \
+    {                                                                                   \
+        uint64_t _api_count_in, _api_count_out;                                         \
+        WT_ACQUIRE_READ(_api_count_out, S2C(s)->counter##_out);                            \
```

Comment:
@luke-pearson - this is the code we were discussing in Slack.
I'm confident the `WT_ACQUIRE_READ` is safe enough for what I want - let me know if it would be better to use a different mechanism.

Reply:
Well this is certainly a case that we haven't yet got a guideline for. I'd say a matching atomic load is still better as this doesn't pair with a `RELEASE_WRITE`. That is, if you write atomically, read atomically. Still SEQ_CST is a heavy hammer but we don't have much else. If it's not performant I'd think then there needs to be a bigger conversation. You'd need to implement a SEQ_CST load macro.

Reply:
cc: @ajmorton

Reply:
I agree with Luke that it would be nice to pair `ACQUIRE_READ`s with `RELEASE_WRITE`s. This is a constraint I'd like to add in the WT-10967 documentation task to make it easier for developers to reason about the code.

We could do
```
__wt_atomic_load64(&S2C(s)->counter##_out);
/* Make sure we read counter_out before counter_in */
WT_ACQUIRE_BARRIER;
__wt_atomic_load64(&S2C(s)->counter##_in);
``` 
which gives us relaxed loads for the atomicity and still maintains ordering.

I need to read through the code changes and prior PR comments, but stepping back for a second I have some general questions:
- `API_COUNTER_REALIZE` is only used in the stat collection thread. Do we need to worry about performance impacts? The stat increments in the API entry/exit are more impactful.
- Why are we using separate `_api_count_in` and `_api_count_out` counters and then deriving the count? A single `api_count` that gets decremented on API exit would do away with any ordering concerns.